### PR TITLE
[Spec] DFlash2: local convolution + candidate selector

### PR DESCRIPTION
**Pull request by SubSir**
## Motivation

Two additions to the DFlash draft, both switched on by the checkpoint. A DFlash
checkpoint declares neither and takes the path it always took.

**Grouped dynamic depthwise convolution** inside each block, so a proposal
position can see the ones before it without another backbone pass.
`out[i,c] = Σ_t (base[t,c] + δ[i,t,g(c)]) · x[i−t,c]`, taps zero across the block
boundary. Wraps each sublayer in and out from one projection of its input.
On when `conv_kernel_size` and `conv_group_size` are set.

**Candidate selector.** Instead of an independent argmax per slot, keep the
target head's top-K per slot, score adjacent transitions
`edge(p→c) = ⟨A[p] ⊙ project(h), B[c]⟩ + unary[c]`, and walk the best path from
the verified anchor. At T>0 it walks by inverse CDF and returns q over the K
candidates for the lossless verify. On when `selector_rank` and
`selector_top_k` are set.

## Modifications

```
srt/models/dflash.py                 the convolution, the selector, the walk kernel
srt/speculative/dflash_utils.py      four config fields
srt/speculative/dflash_worker_v2.py  the proposal and the T>0 verify
```

The selector's decode is folded into the draft CUDA graph, greedy and sampled
alike. The walk is one Triton program per request: a slot's K scores stay in
registers, greedy rows take the argmax and emit a one-hot q, and the slot-to-slot
dependency is a loop inside the program rather than a kernel per slot. No new
server args, no new env vars, nothing outside DFlash touched.

## Accuracy Tests

```
python -m pytest test/registered/unit/spec/test_dflash_logits.py
python -m pytest test/registered/unit/model_executor/model_runner_components/test_spec_aux_hidden_state.py
```

`4 passed` and `5 passed`. `test_dflash_logits.py` covers the selector's sampled path end to end,
the mixed greedy/sampling batch — each row identical to running it alone, with
greedy rows asserted to carry a one-hot q — and the rejection of a quantized
target LM head.

Lossless decoding is the property that matters here and it is structural: greedy
walks emit the point mass at the token they take, and sampled walks return the
realized q over the K candidates, so the verify sees the proposal distribution it
actually drew from.

## Speed Tests and Profiling

Published evaluation for
[`z-lab/Qwen3.8-27B-DFlash2`](https://huggingface.co/z-lab/Qwen3.8-27B-DFlash2)
on `Qwen/Qwen3.8-27B`: SGLang on one H200 with FlashAttention 3, block size 8
(seven draft tokens per verification step), Qwen3.8's recommended sampling
(temperature 1.0, top-p 0.95, top-k 20) at `xhigh` reasoning effort, 4096 max new
tokens, prompts formatted by [`z-lab/dflash`](https://github.com/z-lab/dflash).
Compared against the target's built-in MTP and a community DSpark drafter,
[`RadixArk/Qwen3.8-27B-DSpark`](https://huggingface.co/RadixArk/Qwen3.8-27B-DSpark),
both also proposing seven tokens per step.

### Acceptance length

Per-request mean of completion tokens over verification steps.

| Dataset | MTP | DSpark | DFlash 2 |
|---------|----:|-------:|---------:|
| GSM8K | 5.02 | 4.36 | **5.46** |
| MATH-500 | 4.72 | 3.92 | **5.28** |
| HumanEval | 3.91 | 3.30 | **4.39** |
| MBPP | 3.99 | 3.51 | **4.79** |
| MT-Bench | 3.74 | 3.01 | **4.10** |

### Throughput, GSM8K

Output tokens over end-to-end wall time, with speedup over autoregressive
decoding.

| conc | autoregressive | MTP | DSpark | DFlash 2 |
|------|---------------:|----:|-------:|---------:|
| 1 | 68.9 | 178.5 (2.59×) | 185.3 (2.69×) | **236.1 (3.43×)** |
| 8 | 467.2 | 1,022.1 (2.19×) | 1,040.8 (2.23×) | **1,328.7 (2.84×)** |
| 32 | 1,329.8 | 1,381.1 (1.04×) | 1,506.5 (1.13×) | **1,922.5 (1.45×)** |

Full tables for the other four datasets, and for
[`z-lab/Muse-Glimmer-30B-DFlash2`](https://huggingface.co/z-lab/Muse-Glimmer-30B-DFlash2)
against Meta's published DFlash assistant, are on the model cards.

### What it costs

The convolution and the selector are the whole difference between a DFlash draft
and a DFlash 2 draft, so their cost is what a step loses by having them. Measured
against Meta's published DFlash assistant on the same target,
[`meta-models/Muse-Glimmer-30B`](https://huggingface.co/meta-models/Muse-Glimmer-30B),
block 16, GSM8K, four H200 shards. `cost` is `(1 + accept gain) / (1 + throughput
gain) - 1` — the throughput gain the extra acceptance would have produced had the
step stayed the same length, against the gain it actually produced.

| conc | acceptance | throughput | cost | | acceptance | throughput | cost |
|------|-----------:|-----------:|-----:|---|-----------:|-----------:|-----:|
| | **T=0** | | | | **T=1** | | |
| 1 | +13.6% | +11.7% | 1.7% | | +24.4% | +21.1% | 2.7% |
| 8 | +14.0% | +11.4% | 2.3% | | +24.5% | +21.5% | 2.5% |
| 32 | +14.1% | +11.9% | 2.0% | | +24.0% | +21.2% | 2.3% |
| 64 | +14.0% | +12.3% | 1.5% | | +24.3% | +22.2% | 1.7% |

So 1.5–2.7%, and it falls as the batch grows: the target verify grows with the
batch while the selector's work does not keep pace. It falls as the target grows
for the same reason.

Acceptance enters this table only as a ratio between the two drafts, both
measured the same way — as total accepted tokens over total verification steps,
where the model card reports a mean over per-request ratios. The two statistics
differ on absolute values but nearly cancel in the gain: +24.3% here at
concurrency 64 against the card's +21.0% for the same pair of checkpoints.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #32199057265](https://github.com/sgl-project/sglang/actions/runs/32199057265)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #32199057143](https://github.com/sgl-project/sglang/actions/runs/32199057143)

### CODE REVIEW

**Review by hnyls2002 — APPROVED** [review]
LGTM

**Comment by hnyls2002** [comment]
### CI States

Latest PR Test (Base):  [Run #32190273768](https://github.com/sgl-project/sglang/actions/runs/32190273768)
Latest PR Test (Extra):  [Run #32190273531](https://github.com/sgl-project/sglang/actions/runs/32190273531)
